swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1281 @@
+                        dx[d] = np.random.uniform() * 0.3 * H * sign
+
+                    x[oind] = x[uind] + dx
+                    touched[oind] = True
+                    touched[uind] = True
+                    moved += 1
+                    break
+
+    if moved > 0:
+
+        # check boundary conditions
+        if icSimParams["periodic"]:
+            for d in range(ndim):
+                x[x[:, d] > boxsize[d], d] -= boxsize[d]
+                x[x[:, d] < 0.0, d] += boxsize[d]
+        else:
+            for d in range(ndim):
+
+                # move them away from the edge by a random factor of mean "cell size" boxsize/npart^ndim
+                mask = x[:, d] > boxsize[d]
+                x[mask, d] = boxsize[d] * (
+                    1.0 - np.random.uniform(x[x > 1.0].shape) / npart ** (1.0 / ndim)
+                )
+
+                mask = x[:, d] < 0.0
+                x[mask, d] = (
+                    np.random.uniform(x[mask, d].shape)
+                    * boxsize[d]
+                    / npart ** (1.0 / ndim)
+                )
+
+    return x, indices[touched]
+
+
+def _IC_write_intermediate_output(
+    iteration: int,
+    x: np.ndarray,
+    m: np.ndarray,
+    rho: np.ndarray,
+    h: np.ndarray,
+    icSimParams: dict,
+):
+    r"""
+    Write an intermediate output of current particle positions, densities,
+    masses, and smoothing lengths.
+
+    Parameters
+    ----------------
+
+    iteration: int
+        current iteration number
+
+    x: np.ndarray
+        current particle coordinates
+
+    m: np.ndarray
+        particle masses
+
+    rho: np.ndarray
+        particle densities
+
+    h: np.ndarray
+        particle smoothing lengths
+
+    icSimParams: dict
+        a dict containing simulation parameters as returned by 
+        ``IC_set_IC_params()``.
+
+
+
+    Returns
+    -------------------
+
+    Nothing.
+
+    """
+
+    from swiftsimio import Writer
+
+    nx = icSimParams["nx"]
+    ndim = icSimParams["ndim"]
+    npart = nx ** ndim
+
+    ICunits = unyt.UnitSystem(
+        "IC_generation", icSimParams["unit_l"], icSimParams["unit_m"], unyt.s
+    )
+
+    W = Writer(ICunits, icSimParams["boxsize"])
+    W.gas.coordinates = unyt_array(x, icSimParams["unit_l"])
+    W.gas.smoothing_length = unyt_array(h, icSimParams["unit_l"])
+    W.gas.masses = unyt_array(m, icSimParams["unit_m"])
+    W.gas.densities = unyt_array(
+        rho, W.gas.masses.units / W.gas.coordinates.units ** ndim
+    )
+
+    # invent some junk to fill up necessary arrays
+    W.gas.internal_energy = unyt_array(
+        np.zeros(npart, dtype=np.float), unyt.m ** 2 / unyt.s ** 2
+    )
+    W.gas.velocities = unyt_array(np.zeros(npart, dtype=np.float), unyt.m / unyt.s)
+
+    # If IDs are not present, this automatically generates
+    fname = "IC-generation-iteration-" + str(iteration).zfill(5) + ".hdf5"
+    W.write(fname)
+
+    return
+
+
+def _IC_plot_current_situation(
```

Comment:
We should at the very least move this to a separate file. 

I don't like having underscores here, that implies they are private and unsed internally rather than for debugging. We can create `swiftsimio/initial_conditions/debug.py` and put this there. Or `swiftsimio/initial_conditions/plotting.py`.

Reply:
I meant debugging _before we merge this_. I initially intended to remove it fully at the end, but with the intermediate data dumps, i could just use it for a script/example to quickly just plot these or something. Or we might scratch it anyway. Anyhow, I moved it to its own file now.